### ISSUE TITLE
Fix Broken TypeScript Merge, Publish `@next` tag and Switch `main` back to `master`

### ISSUE THREAD

**Issue by thesunny** [ianstormtaylor/slate#4003]
#### Do you want to request a _feature_ or report a _bug_?

Fix a broken merge into `main` from this TypeScript PR https://github.com/ianstormtaylor/slate/pull/3835

#### What's the current behavior?

For posterity and as a post mortem, this describes the steps that led to the broken merge (I take full responsibility), the steps required to resolve them including a description of how we will prevent this from happening in the future.

The original goal was to (a) accept this TypeScript PR https://github.com/ianstormtaylor/slate/pull/3835 and (b) publish it to npm with an `@next` tag as per this proposal https://github.com/ianstormtaylor/slate/issues/3975

What happened
- I checked the PR to ensure that it was passing CI. It was not passing due to a Netlify issue which was resolved with @ianstormtaylor.
- I accepted the PR through the GitHub UI
- I cloned `slate` into a clean directory
- I upgraded the versions to the next `minor` versions using lerna and made a few script changes to help with deploying to `@next`
- I pushed those changes to `origin/main` (GitHub)
- I published to npm adding the `@next` tags which completed successfully
- Then...
- Later, due to a question from @BrentFarese asking which branch I made the release from, I realized that the `main` branch I published with an `@next` tag did not contain the TypeScript PR. When I accepted the PR, GitHub merged the PR into the `master` branch and not the `main` branch because at the time the PR was created, `master` was the default branch.
- In order to fix this, I merged `main` into `master`, pushed back to `origin/main` (GitHub) and then published with the `@next` tag
- This was a mistake. The PR was passing CI because it was against the old default `master` branch. When merged into `main`, the PR was in fact broken. The GitHub UI does not show this because the PR against `master` was okay. The PR broke after the merge because there were a number of newly accepted PRs into `main` after the TypeScript PR was created off of `master`.

At this point, the merge was pushed to GitHub so it's not easy to take it back. It's also not trivial to fix because it was a big TypeScript PR that hit a lot of code. If a reverse commit is used, it may become difficult to merge the TypeScript PR into the branch again because, technically, it has already been applied. Also, I've personally never done a reverse commit so I don't want to do it because I fear making a bad situation worse.

Also, as a note for posterity, I pushed changes to GitHub before publishing to NPM. Andrew Herron (not sure his GitHub username) has made me aware that I can publish without pushing which is safer. This would have saved us from having a broken `main` branch.

#### What's the expected behavior?

There are two issues to resolve

1. Fix the TypeScript types and publish again using an `@next` tag
2. Set the default branch back to `master` to prevent this from happening again. When GitHub tooling helps us move the PRs to `main`, we can then discuss switching the default branch back to `master`. Note that this has been discussed and @ianstormtaylor has approved of the change https://slate-js.slack.com/archives/CC58ZGGU9/p1606312906203400

Since there are a number of things going on, I think the simplest is to fix each of the issues in isolation from each other. For this reason, I think we don’t want to switch back to master before getting the release working properly on main. The approach is:

1. Create a branch from main that those who worked on the TypeScript types can fix the types on. This has already been created here https://github.com/ianstormtaylor/slate/tree/fix-typescript-merge as the `fix-typescript-merge` branch
2. Once those fixes pass all tests and CI, merge that back into main
3. npm publish from main using the @next tag
4. merge main back into master
5. Set master as the default

May I (politely) ask those who worked on the TypeScript PR to help fix the typing issue? @timbuckley @BrentFarese @mdmjg

**Comment by adamhaeger** [comment]
Im guessing this is the reason the CustomTypes aren't included when you install 0.60.0? 

If so, it might be a good idea to state that in a comment here: https://github.com/ianstormtaylor/slate/issues/3725 as thats the issue that comes up when googling :)

**Comment by adamhaeger** [comment]
Hi, thanks for the great work on slate! Any idea when this one will be fixed? Really need the types feature :)

**Comment by thesunny** [comment]
The TypeScript fixes from @BrentFarese have been merged. Thank you Brent for the contribution.

The fixes are published under the `@next` tag.

I've also added documentation on publishing an `@next` release to `Contributing.md`.

We still need to switch the default branch back to `master`.

As I don't have permission to do this, can you switch this back @ianstormtaylor after merging `main` into `master`? Alternatively, add that permission to my username and I can do it. Thanks!

Once that is complete, this issue can be closed.

**Comment by thesunny** [comment]
Default branch has been switch back to `master` by Ian. Closing this issue.
